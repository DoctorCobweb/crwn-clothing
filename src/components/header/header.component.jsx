import React from 'react';
import { Link } from 'react-router-dom';
import { connect } from 'react-redux';
import { createStructuredSelector } from 'reselect';

import { auth } from '../../firebase/firebase.utils';
import CartIcon from '../../components/cart-icon/cart-icon.component';
import CartDropdown from '../../components/cart-dropdown/cart-dropdown.component';
import { selectCartHidden } from '../../redux/cart/cart.selectors';
import { selectCurrentUser } from '../../redux/user/user.selector';

// special syntax for svg importing to make a **component**
// https://facebook.github.io/create-react-app/docs/adding-images-fonts-and-files 
// the ReactComponent import name is special and tells Create React App (CRA)
// that you want a React component that renders an SVG, rather that its filename.
// nb. this is a CRA special syntax
import { ReactComponent as Logo } from '../../assets/crown.svg';
import './header.styles.scss';

const Header = ({ currentUser, hidden }) => (
  <div className="header">
    <Link className="logo-container" to="/">
      <Logo className="logo" />
    </Link>
    <div className="options">
      <Link className="option" to="/shop">
        SHOP
      </Link>
      <Link className="option" to="/contact">
        CONTACT 
      </Link>
      {
        currentUser ?
        <div className='option' onClick={() => auth.signOut()}>SIGN OUT</div>
        :
        <Link className='option' to='/signin'>SIGN IN</Link>
      }
      <CartIcon />
    </div>
    {
      hidden ? null : <CartDropdown />
    }
  </div>
);

const mapStateToProps = createStructuredSelector({
  currentUser: selectCurrentUser,
  hidden: selectCartHidden
});

export default connect(mapStateToProps)(Header);