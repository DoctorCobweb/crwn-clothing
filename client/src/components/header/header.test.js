import React from 'react';
import { shallow } from 'enzyme';

import { Header } from './header.component';
import CartDropdown from '../cart-dropdown/cart-dropdown.component';


describe('Header component', () => {
  let wrapper;
  let mockCurrentUser;
  let mockSignOutStart;
  let mockHidden;

  beforeEach(() => {
    mockSignOutStart = jest.fn();
    mockHidden = true;
    mockCurrentUser = {
      uid: '123',
    };

    const mockProps = {
      currentUser: mockCurrentUser,
      signOutStart: mockSignOutStart,
      hidden: mockHidden,
    }

    wrapper = shallow(<Header {...mockProps}/>);
  });

  it('should render Header component', () => {
    expect(wrapper.debug()).toMatchSnapshot();
  });

  describe('if currentUser is present', () => {
    it('should render sign out link', () => {
      expect(wrapper.find('OptionLink').at(2).text()).toBe('SIGN OUT');
    });

    it('should call signOutStart when SIGN OUT link is clicked', () => {
      wrapper.find('OptionLink').at(2).simulate('click');
      expect(mockSignOutStart).toHaveBeenCalled();
    });
  })

  describe('if currentUser is null', () => {
    it('should render sign in link', () => {
      const newMockProps = {
        currentUser: null,
        signOutStart: mockSignOutStart,
        hidden: true,
      };
      const newWrapper = shallow(<Header {...newMockProps}/>);
      expect(newWrapper.find('OptionLink').at(2).text()).toBe('SIGN IN');
    });
  });

  describe('if hidden is true', () => {
    it('should not render CartDropdown', () => {
      expect(wrapper.exists(CartDropdown)).toBe(false);
    });
  });

  describe('if hidden is false', () => {
      const newMockProps = {
        currentUser: null,
        signOutStart: mockSignOutStart,
        hidden: false,
      };
      const newWrapper = shallow(<Header {...newMockProps}/>)
    it('should render CartDropdown', () => {
      expect(newWrapper.exists(CartDropdown)).toBe(true);
    });
  });
});