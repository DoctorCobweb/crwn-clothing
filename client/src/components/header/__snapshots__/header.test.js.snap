// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Header component should render Header component 1`] = `
"<HeaderContainer>
  <LogoContainer to=\\"/\\">
    <ForwardRef(SvgCrown) className=\\"logo\\" />
  </LogoContainer>
  <OptionsContainer>
    <OptionLink to=\\"/shop\\">
      SHOP
    </OptionLink>
    <OptionLink to=\\"/contact\\">
      CONTACT
    </OptionLink>
    <OptionLink as=\\"div\\" onClick={[Function: mockConstructor]}>
      SIGN OUT
    </OptionLink>
    <Connect(CartIcon) />
  </OptionsContainer>
</HeaderContainer>"
`;
